enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1090 @@
+  unsupported by OpenShift)
+* Management of the etcd cluster state (the etcd operator will handle this separately)
+* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)
+* Management of anything that falls out of the scope of Machine API
+
+## Proposal
+
+A new CRD `ControlPlaneMachineSet` will be introduced into OpenShift and a respective `control-plane-machine-set-operator`
+will be introduced as a new second level operator within OpenShift to perform the operations described in
+this proposal.
+
+The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
+The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
+cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
+of updated configuration to the Control Plane Machines.
+
+### User Stories
+
+- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
+  Machines (eg those that have been removed by the cloud provider, or those failing MachineHealthChecks)
+- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
+  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
+- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
+  Machines such that I can test changes with a single replica before applying the change to all replicas
+- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
+  scale down my control plane before adding new Control Plane Machines with newer configuration, notably, my
+  environment does not have capacity to add additional Machines during updates
+
+### API Extensions
+
+We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1b` API group. It will be based
+on the spec and status structures defined below.
+
+```go
+// ControlPlaneMachineSet represents the configuration of the ControlPlaneMachineSet.
+type ControlPlaneMachineSetSpec struct {
+	// Replicas defines how many Control Plane Machines should be
+	// created by this ControlPlaneMachineSet.
+	// This field is immutable and cannot be changed after cluster
+	// installation.
+	// The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
+	// 3 and 5 are the only valid values for this field.
+	// +kubebuilder:validation:Enum:=3;5
+	// +kubebuilder:default:=3
+	// +kubebuilder:validation:Required
+	Replicas *int32 `json:"replicas"`
```

Comment:
Why is this a pointer if it is a required field?

Reply:
I believe there were some comments on this before, IIRC this is because we are using a scale subresource and with other resources that use the scale subresoruce, the spec replicas are expected to be a pointer. To ensure consistency between this and other scalable resources, we kept this one as a pointer

Reply:
Yeah this is 100% a future proofing action. It's likely we will want to support scale in the future from conversations I have had, but for now we aren't going to. The ill-effect of having this as a pointer is minimal given the issues we would face if we wanted to change it from a reference to a pointer later IMO, so better to keep as a pointer for now to be consistent and future proof ourselves

Reply:
The MachineSpec is within the OpenShiftMachineV1Beta1Template in this API, we include a `machinev1beta1.MachineSpec` as `spec` which is the template of the Machine we use to stamp out new Machines. 

If the user changes the providerSpec, eg the vmSize, we will perform a roll out for them